RHDEVDOCS 3306: Document running image build tasks as unprivileged bu… (#50364)
openshift/openshift-docs: modules/op-running-as-root-user-namespace.adoc

.. Get the name of the node where the buidah container is running:
+
[source,terminal]
----
$ oc get pod -o wide
----

.. Run a debugging pod on the node where the `buildah` pod is running:
+
[source,terminal]
----
$ oc debug nodes/<node-name>
----

.. Change the root directory for the current running processes:
+
[source,terminal]
----
$ chroot /host
----

.. List the user namespaces:
+
[source,terminal]
----
$ lsns -t user
----

[NOTE]
====
* Although {pipelines-shortname} runs your `buildah` container as a root user, CRI-O runs the container as a user in the host namespace.

* When you run a command in a user namespace, your container runs as root (user id 0) but has user privileges on the host. {product-title} uses CRI-O as its container runtime interface. CRI-O has a workload feature to run a pod in the user namespace.
====

.Example: Modified Buildah task
[source,yaml]
----
...
apiVersion: tekton.dev/v1beta1
kind: Task
metadata:
  name: buildah-userns
  labels:
    app.kubernetes.io/version: "0.1"
  annotations:
    tekton.dev/pipelines.minVersion: "0.12.1"

>     tekton.dev/tags: image-build

Reviewer: The annotations that were mentioned earlier in this doc are missing here

    Dockerfile to assemble a container image, then pushes that image to a
    container registry.

  params:
  - name: IMAGE
    description: Reference of the image buildah will produce.
  - name: BUILDER_IMAGE
    description: The location of the buildah builder image.
    default: registry.redhat.io/rhel8/buildah@sha256:99cae35f40c7ec050fed3765b2b27e0b8bbea2aa2da7c16408e2ca13c60ff8ee
  - name: STORAGE_DRIVER
    description: Set buildah storage driver
    default: vfs
  - name: DOCKERFILE
    description: Path to the Dockerfile to build.
    default: ./Dockerfile
  - name: CONTEXT
    description: Path to the directory to use as context.
    default: .
  - name: TLSVERIFY
    description: Verify the TLS on the registry endpoint (for push/pull to a non-TLS registry)
    default: "true"
  - name: FORMAT
    description: The format of the built container, oci or docker
    default: "oci"
  - name: BUILD_EXTRA_ARGS
    description: Extra parameters passed for the build command when building images.
    default: ""
  - name: PUSH_EXTRA_ARGS
    description: Extra parameters passed for the push command when pushing images.
    type: string

>     default: ""

Reviewer: Task in 1.8.0 has one more param and uses newer image, maybe we should updated the whole code sample
```
  - default: "false"
    description: Skip pushing the built image
    name: SKIP_PUSH
    type: string
```

Author: I will apply these patches in a separate task: https://issues.redhat.com/browse/RHDEVDOCS-4657
For the time being, I am in favor of closing the issue so that at least docs till 1.7 are present.

Author: Based on discussion in the thread https://coreos.slack.com/archives/CG5GV6CJD/p1667818845838809, the content will be completely rewritten, and therefore these comments are not useful anymore.

  workspaces:
  - name: source

  results:
  - name: IMAGE_DIGEST
    description: Digest of the image just built.

  steps:
  - name: build
    securityContext:

>       runAsUser: 1000

Reviewer: task in 1.8.0 has completely different `securityContext`, only the other doc mentions `runAsUser: 1000`

Author: I will apply these patches in a separate task: https://issues.redhat.com/browse/RHDEVDOCS-4657
For the time being, I am in favor of closing the issue so that at least docs till 1.7 are present.

Author: Based on discussion in the thread https://coreos.slack.com/archives/CG5GV6CJD/p1667818845838809, the content will be completely rewritten, and therefore these comments are not useful anymore.

// Module included in the following assemblies:
//
// * cicd/pipelines/running-workloads-and-buildah-as-user-namespaces-on-openshift-pipelines.adoc
:_content-type: PROCEDURE

[id="op-running-as-root-user-namespace_{context}"]
= Running buildah as root in a user namespace

> If you are using {pipelines-shortname} to run pods in the user namespace, you must pass the annotations that the CRI-O container engine expects. You can add these annotations for all users, or for a specific namespace.

Reviewer 2: ```suggestion
If you use {pipelines-shortname} to run pods in the user namespace, you must pass the annotations that the CRI-O container engine expects. You can add these annotations for all users or for a specific namespace.
```
- change to simple present tense for concision
- omit unnecessary comma

> 

Reviewer 2: Are there any prerequisites for this procedure? If so, list them in a `.Prerequisites` section before the `.Procedure` section

Author: Modified the existing introduction, but I think a formal prerequisite for explicitly stating "logged in as a cluster administrator" would be an overkill. The requirement of being logged in as a cluster administrator is so pervasive and implied in almost everything that it is hardly ever stated explicitly.
I can't think of any other prerequisite.

Reviewer 2: Procedures in the OCP monitoring docs always explicitly state what user role is required to do the procedure, whether a role with admin or just edit permissions, so in my experience it's not out of the ordinary to state it. However, if your area is only ever touched by users with admin rights, then I suppose it makes sense.

.Procedure

. Export the `buildah` cluster task as a task to the project's namepsace.
+
[source,terminal]
----
$ tkn task create --from=buildah
----

. Edit the `buildah` task within the project's namespace:
+
[source,terminal]
----
$ oc edit task buildah
----

. Add annotations to the task:
+
[source,yaml]
----
...
annotations:
  io.kubernetes.cri-o.userns-mode: "auto"
  io.openshift.builder: "true"
...
----

> . Verify that the `buildah` task's container is running as root in the user-defined namespace. 

Reviewer 3: user doesn't have to go on the node to activate non-root, it's ony useful if as an admin you need to make sure the setup works prroperly

.. Get the name of the node where the buidah container is running:
+
[source,terminal]
----
$ oc get pod -o wide
----

.. Run a debugging pod on the node where the `buildah` pod is running:
+
[source,terminal]
----
$ oc debug <node-name>
----

> .. Create a separate virtualized copy of your host environment:

Reviewer 3: this is a weird definition of chroot, i think it needs some rephrasing to don't make it confusing...

+
[source,terminal]
----
$ chroot /host
----

.. List the user namespaces:
+
[source,terminal]
----
$ lsns -t user
----

[NOTE]
====
* Although {pipelines-shortname} runs your `buildah` container as a root user, CRI-O runs the container as a user in the host namespace.

* When you run a command in a user namespace, your container runs as root (user id 0) but has user privileges on the host. {product-title} uses CRI-O as its container runtime interface. CRI-O has a workload feature to run a pod in the user namespace.
====

> .Example: Modified Buildah task

Reviewer 4: This task is wrong for this document : here we are documenting how to run as root in the container and user on the host (so user-ns). But the Task ask to run as a specificy `1000` user which doesn't make sense imho, as we are explicitely not running as root in the container then.

> You can use the following `buildah` task configuration for reference:

Reviewer 2: ```suggestion
```
Consider removing this intro sentence for minimalism. It's not adding any new content that's not already implied by the `.Example` heading -- that is, because you have the heading `.Example: Modified task`, the user knows the sample is meant as a reference.

> +

Reviewer 2: ```suggestion

```
Remove extra `+` to fix the `.Example` heading not rendering properly
